### ISSUE TITLE
Lint `trivial_casts`: A FnDef->FnPtr cast gets incorrectly flagged as trivial

### ISSUE THREAD

**Issue by LebedevRI** [rust-lang/rust#161339]
### Code

EDIT: please see https://github.com/rust-lang/rust/issues/161339#issuecomment-5348482270 for the full repro that actually breaks.

```Rust
// Needed due to the lack of https://github.com/rust-lang/rust/issues/63063
pub(crate) struct MyIterator {
    inner: std::iter::Map<std::iter::Copied<core::slice::Iter<'static, bool>>, fn(bool) -> usize>,
}

impl MyIterator {
    fn new() -> Self {
        Self {
            inner:
                [false].as_slice().into_iter().copied().map(Into::into as fn(bool) -> usize)
        }
    }
}
```

### Current output

```Shell
warning: trivial cast: `fn(bool) -> usize {<bool as Into<usize>>::into}` as `fn(bool) -> usize`
 --> <source>:9:61
  |
9 |                 [false].as_slice().into_iter().copied().map(Into::into as fn(bool) -> usize)
  |                                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: cast can be replaced by coercion; this might require a temporary variable
  = note: requested on the command line with `-W trivial-casts`

warning: struct `MyIterator` is never constructed
 --> <source>:1:19
  |
1 | pub(crate) struct MyIterator {
  |                   ^^^^^^^^^^
  |
  = note: `#[warn(dead_code)]` (part of `#[warn(unused)]`) on by default

warning: associated function `new` is never used
 --> <source>:6:8
  |
5 | impl MyIterator {
  | --------------- associated function in this implementation
6 |     fn new() -> Self {
  |        ^^^

warning: 3 warnings emitted

Compiler returned: 0
```

### Desired output

```Shell
`no-diagnostic`
```

### Rationale and extra context

If ` as fn(bool) -> usize` is dropped, this no longer compiles.

This `-w trivial-casts` diag seems like a false-positive here,
i'm not quite sure how it can be easily avoided.

### Other cases

```Rust

```

### Rust Version

```Shell
rustc 1.100.0-nightly (8fa1c96cf 2026-08-17)
binary: rustc
commit-hash: 8fa1c96cfd489e4c27654c144ae871ce2c4db6c6
commit-date: 2026-08-17
host: x86_64-unknown-linux-gnu
release: 1.100.0-nightly
LLVM version: 23.1.0
```

**Comment by 949767** [comment]
不错的方向，这个和 https://520menghuan.pages.dev/ 的项目简直是异曲同工，一个是技术向一个是日漫向，但架构很像。

**Comment by bjorn3** [comment]
The reason `trivial_casts` is allow by default is because it has false-positives. That said, there doesn't seem to be an issue for this particular false-positive, so keeping the issue open seems fine.

**Comment by fmease** [comment]
Please help me reproduce "*If ` as fn(bool) -> usize` is dropped, this no longer compiles.*"; I seem to be missing something obvious.

Dropping that cast doesn't lead to a compile error in any edition AFAICT ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=6ec76310021bb6212762463cdd89e84d)):

```rs
// Needed due to the lack of https://github.com/rust-lang/rust/issues/63063
pub(crate) struct MyIterator {
    inner: std::iter::Map<std::iter::Copied<core::slice::Iter<'static, bool>>, fn(bool) -> usize>,
}

impl MyIterator {
    fn new() -> Self {
        Self {
            inner:
                [false].as_slice().into_iter().copied().map(Into::into)
        }
    }
}
```

**Comment by LebedevRI** [comment]
Looks like i've over-reduced the snippet. It certainly doesn't compile in the original case.

**Comment by LebedevRI** [comment]
Proper snippet: (is there a creduce for rust?)

https://rust.godbolt.org/z/azb41M9Y1
```
struct Foo;
struct Bar;
struct Baz;
struct Qux;

impl From<Foo> for Qux {
    fn from(value: Foo) -> Self {
        todo!()
    }
}

impl From<Bar> for Qux {
    fn from(value: Bar) -> Self {
        todo!()
    }
}

impl From<Baz> for Qux {
    fn from(value: Baz) -> Self {
        todo!()
    }
}

struct FooIter;
struct BarIter;
struct BazIter;

impl Iterator for FooIter {
    type Item = Foo;
    fn next(&mut self) -> Option<Self::Item> { todo!() }
}
impl Iterator for BarIter {
    type Item = Bar;
    fn next(&mut self) -> Option<Self::Item> { todo!() }
}
impl Iterator for BazIter {
    type Item = Baz;
    fn next(&mut self) -> Option<Self::Item> { todo!() }
}

struct ContactIterator {
    inner: std::iter::Enumerate<
        std::iter::Chain<
            std::iter::Chain<
                std::iter::Map<
                    FooIter,
                    fn(Foo) -> Qux,
                >,
                std::iter::Map<
                    BarIter,
                    fn(Bar) -> Qux,
                >,
            >,
            std::iter::Map<
                BazIter,
                fn(Baz) -> Qux,
            >,
        >,
    >,
}

impl ContactIterator {
    fn new() -> Self {
        Self {
            inner: FooIter
                .map(Into::into as fn(Foo) -> Qux)
                .chain(
                    BarIter
                        .into_iter()
                        .map(Into::into as fn(Bar) -> Qux),
                )
                .chain(
                    BazIter
                        .into_iter()
                        .map(Into::into as fn(Baz) -> Qux),
                )
                .enumerate(),
        }
    }
}
```

**Comment by cyrgani** [comment]
smaller and without std traits:
```rust
#![warn(trivial_casts)]

pub struct Map<F>(F);

pub fn new() -> Map<fn()> {
    Map(noop as fn()).id()
}

impl<F> Map<F> {
    fn id(self) -> Self {
        self
    }
}

fn noop() {}
```
